### DISCUSSION TITLE
Additional Switching Station URLs Problems

### DISCUSSION THREAD

**Discussion by waszee** [pschatzmann/arduino-audio-tools#1782] [Q&A] [answered]
I am having trouble switching between the following 4 URL internet radio streams. Using a  Audio.h version I am able to connect to host and listen to stations per index. Using AudioTools with the player to switch the stations by index is not working well.  Stations 0 to 1 switch fine and play back.  Moving on from station 1 [Scenic Route]  to station 2 [Nashville], playback fails and then after a pause the player will actually connect to station 3 [Irish Folk]. Moving on to station 3 from 2, will connect correctly to Irish Folk but has gaps in playback sound.  If I move backwards to station 2 [the German Nashville sound station], now connects.  Playback has noticeable sound gaps.  The problem appears to be something in the station streams or handling of addresses.  Other urls such as station 0 [Folk Alley]  and station 1 [Scenic Route] play well and switch well.    I am attaching the Serial logs of metadata callback for the audio and player versions for comparison.  The version using audio.h with audio info output has more info details about the streams but I am unable to pick out the problem.  They all seem to have 44100 bits per second, 2 channel, and metaint 16000 data.  Hope somebody has some suggestions.

`const char *urls[] = {
  "https://freshgrass.streamguys1.com/folkalley-128mp3-tunein",
  "https://samcloud.spacial.com/api/listen?sid=86781&amp;m=sc&amp;rid=282728&amp;t=ssl",
  "https://nashville.stream.laut.fm/nashville",
  "https://irish-folk.stream.laut.fm/irish-folk",
  "https://rt1.radio/ycg57",
...`

[device logs_audiotools.txt](https://github.com/user-attachments/files/17689108/device.logs_audiotools.txt)
[device logs_audio_ver.txt](https://github.com/user-attachments/files/17689109/device.logs_audio_ver.txt)

**Comment by pschatzmann** [comment, marked as answer]
I have no clue what you talk about: what is audio.h ?
What logic are you using to select a channel ?
I think you need to make some more test cases to nail down the problem...

Also try to avoid https if plain vanilla http is working: https adds a lot of overhead!

**Comment by waszee** [comment]
Audio.h is a common ESP32 library from Espressif.  Most of the Arduino IDE examples for I2S use it and I assumed you were very familar with it.   In my case the stations urls are listed in a file on my SD Card.  That list is then read into a string array in the program and the index position used to connect to the host stations.   Below is the audio version.  The player version did not use the SD card and puts the url list in a char * array instead of a string array.    
  
`//**********************************************************************************************************
//*    audioI2S-- I2S   for ESP32,                                                              *
//**********************************************************************************************************
//
// first release on 11/2018
// Version 3  , Jul.02/2020
//
//
// THE SOFTWARE IS PROVIDED "AS IS" FOR PRIVATE USE ONLY, IT IS NOT FOR COMMERCIAL USE IN WHOLE OR PART OR CONCEPT.
// FOR PERSONAL USE IT IS SUPPLIED WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
// WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHOR
// OR COPYRIGHT HOLDER BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE
//

#include <Arduino.h>
#include <Audio.h>
#include "SD_MMC.h"
#include <IRremoteESP8266.h>
#include <IRrecv.h>
#include <IRutils.h>
#include <WiFiMulti.h>
#include <Wire.h>
#include <SSD1306Wire.h>        // legacy: #include "SSD1306.h"

void open_new_song(String filename);
void display_music();
void logoshow(void);
void lcd_text(String text);
void print_song_info();
void print_song_time();
void display_StationVolume();

void check_serial_radio();
void check_vol_change();
void check_station_change();
void check_MP3_play();
void check_control_btn();
int get_music_list(fs::FS &fs, const char *dirname, uint8_t levels, String wavlist[100]);
int get_station_list(fs::FS &fs, const char *dirname, uint8_t levels, String stations[100]);
void listDir(fs::FS &fs, const char * dirname, uint8_t levels);
void get_netname_password();
void change_station();


WiFiMulti wifiMulti;
#define MAX_LINES 100 // sets the max number of stations to keep arrays manageable

SSD1306Wire display(0x3c, 16, 17);   // ADDRESS, SDA, SCL  
//set IR receiver
const uint16_t kRecvPin = 14;  //using pin 14 for ir input data
IRrecv irrecv(kRecvPin);
decode_results results;

//SD MMC Lines
#define SD_MMC_CMD 38 //Please do not modify it.
#define SD_MMC_CLK 39 //Please do not modify it. 
#define SD_MMC_D0  40 //Please do not modify it.

// Some Pins for buttong Toggles

const int Pin_vol_up = 47; // cycles volume in steps of 4 from 1 to 20                                                                                                                                                                                                                                                               
const int Pin_vol_down = 45; //pin48 toggles led on board pin19 and pin20 also had problems
const int Pin_mute = 21; // toggles between v=1 and v=10

const int Pin_down = 1;
const int Pin_pause = 2;
const int Pin_up = 41;
const int Pin_up10 =42;


// i2s audio pins
#define I2S_DOUT       7
#define I2S_BCLK       5
#define I2S_LRC        6

Audio audio;

struct Music_info
{
    String name;
    int length;
    int runtime;
    int volume;
    int status;
    int mute_volume;
} music_info = {"", 0, 0, 0, 0, 0};
 
String song_list[100];
int max_songs = 0;
int position = 0;

String wifi =     "wifi";
String password = "password";
char buf[120];
char buf_sta_vol[120];
String astr ="";
int dvol = 10;

int dstation=0;
const char * dinfo = "";
int maxstation=MAX_LINES;

bool bstop=0;
String stations[MAX_LINES];
/* ={  
  "https://freshgrass.streamguys1.com/folkalley-128mp3-tunein",
  "https://samcloud.spacial.com/api/listen?sid=86781&amp;m=sc&amp;rid=282728&amp;t=ssl",
  "al-mp3-192-9067610", //44
  ... see URL_LIST.TXT file on SDMMC card
 
};
*/
int playMP3=0; // sets to use the SD card for mp3 else go to internet

void get_netname_password(){
  String str;
  File offline = SD_MMC.open("/NET/myNet.txt", FILE_READ);
  while (offline.available()) {
    str = offline.readStringUntil('\n');// str txt: wifi,password
    Serial.println(str);
    int delim = str.indexOf(',');
    String neti=str.substring(0,delim); 
    String pass=str.substring(delim+1);
    Serial.printf("neti: %s \n pass: %s \n", neti,pass);
    wifi=neti;
    password=pass;
  }
  offline.close();
}

void setup() {
  Serial.begin(115200);
  //Wire.begin(16,17);
  //init the lcd display
  display.init();
  display.flipScreenVertically();
  display.setFont(ArialMT_Plain_16);
  display.setTextAlignment(TEXT_ALIGN_LEFT);
  logoshow();

    SD_MMC.setPins(SD_MMC_CLK, SD_MMC_CMD, SD_MMC_D0);
    if (!SD_MMC.begin("/sdcard", true, true, SDMMC_FREQ_DEFAULT, 5)) {
        Serial.println("Card Mount Failed");
        playMP3=0;
    }
    else {
        uint8_t cardType = SD_MMC.cardType();
        if(cardType == CARD_NONE){
            Serial.println("No SD_MMC card attached");
            return;
        }

        Serial.print("SD_MMC Card Type: ");
        if(cardType == CARD_MMC){
            Serial.println("MMC");
        } else if(cardType == CARD_SD){
            Serial.println("SDSC");
        } else if(cardType == CARD_SDHC){
            Serial.println("SDHC");
        } else {
            Serial.println("UNKNOWN");
        }

        uint64_t cardSize = SD_MMC.cardSize() / (1024 * 1024);
        Serial.printf("SD_MMC Card Size: %lluMB\n", cardSize);

        //listDir(SD_MMC, "/", 0);
        lcd_text("SD_MMC OK");
        //Read SD and set up a list of mp3 files to play
        max_songs = get_music_list(SD_MMC, "/mp3", 0, song_list);
        Serial.print("Music file count:");
        Serial.println(max_songs);
        Serial.println("All music:");
        for (int i = 0; i < max_songs; i++)
        {
            Serial.println(song_list[i]);
        }

    }
    // set some pins for button controls to change index or volume
    pinMode(Pin_down, INPUT_PULLUP);
    pinMode(Pin_up,INPUT_PULLUP);
    pinMode(Pin_up10,INPUT_PULLUP);
    pinMode(Pin_vol_up,INPUT_PULLUP);
    pinMode(Pin_vol_down,INPUT_PULLUP);
    pinMode(Pin_mute,INPUT_PULLUP); // might also pause and play toggle

   
    irrecv.enableIRIn();  // enable the IR receiver
  
    //set up the i2s pins for audio out
    audio.setPinout(I2S_BCLK, I2S_LRC, I2S_DOUT);
    audio.setVolume(dvol); // 0...21
    if (playMP3==1) {
        open_new_song(song_list[position]);
        print_song_time();
    }
    else {
      // get the network name and password from file on SDMMC card
        get_netname_password();
        wifi.trim();
        password.trim();
        //Read SD and set up a list of mp3 files to play
       // max_songs = get_music_list(SD_MMC, "/mp3", 0, song_list);
        maxstation= get_station_list(SD_MMC, "/URLs",0,stations);
        //connect to WIFI and Then the HOST STREAM
        WiFi.mode(WIFI_STA);
        wifiMulti.addAP(wifi.c_str(), password.c_str());
        wifiMulti.run();
        if(WiFi.status() != WL_CONNECTED){
            WiFi.disconnect(true);
            wifiMulti.run();
        }
       // maxstation = sizeof(stations)/sizeof(stations[0]);
        Serial.printf("max stations = %d \n",maxstation);
        dstation=0; // starting with first station in the list 
        change_station();                                                                                                                                                                
        //astr=stations[dstation]; astr.trim();
       // Serial.println(astr);
       // astr.toCharArray(buf, sizeof(buf));
       // audio.connecttohost(buf); // default station for setup
        //audio.connecttohost(astr.c_str());
        
        Serial.printf("number of stations %i \n", maxstation);
        Serial.printf("Station %d Vol= %d \n",dstation,dvol);
        
        display_StationVolume();
    }
}

uint run_time = 0;
uint button_time = 0;
int thevolume;
int thestation;
void loop(){
    thevolume=dvol;
    thestation=dstation;
    audio.loop();
    if (irrecv.decode(&results)) {
        check_vol_change();
        check_station_change();
        irrecv.resume();  // Receive the next value
    }
    check_control_btn();
    if (playMP3==1){
        if(thestation!=dstation){
          open_new_song(song_list[dstation]);
        }
/* 
        if (millis() - run_time > 1000) {
            run_time = millis();
            print_song_time();
            display_music();
        }
    
        if (millis() - button_time > 300) {
            //Button logic
            // if next button push advance to next song if available
            if (digitalRead(Pin_up) == 0 ) {
                if (position< max_songs- 1) {
                    position++;
                    Serial.printf("Playing  next song: #%d \n",position);
                    open_new_song(song_list[dstation]);
                    print_song_time();
                    button_time = millis();
                }
                else {
                    position = 0;
                }
            }

*/   
      // if song has finished play next song in list    
      if (audio.isRunning() == false) {
          if (dstation < max_songs - 1) {
              dstation++;
          }
          else {
              dstation=0;
          }
          open_new_song(song_list[dstation]);
      }
    }
    else {
       // if not playing a mp3 list then go to internet url radio stations, i.e. playMP3=0
        check_serial_radio();
        if (thestation!=dstation){
          change_station();
        }
    } 
    
    check_MP3_play(); 

//update to selected station and volume values if they were changed in the checks operations
  if (thevolume!=dvol){
    audio.setVolume(dvol); // 0...21
    display_StationVolume();

  }

}

//////////////////////////////////////////////////////////////
void check_control_btn(){
   
    // scan the control pins for changes
    if (digitalRead(Pin_up)==0) {
       dstation++;
       if (dstation>=maxstation){ dstation=0;}
    }
    if (digitalRead(Pin_up10)==0) {
       dstation+=10;
       if (dstation>=maxstation){ dstation=0;}
    }
    if (digitalRead(Pin_down)==0) {
      dstation--;
        if (dstation<0){ dstation=maxstation-1;}
    }
    if (digitalRead(Pin_vol_up)==0){
        dvol+=4;
        if(dvol>=21){dvol=1;}
    }
    if (digitalRead(Pin_vol_down)==0){
        dvol-=4;
        if(dvol<1){dvol=1;}
    } 
    if (digitalRead(Pin_mute)==0){
        //Toggles volume between 1(mute) and 10 (normal) 
        if (dvol==1){ dvol=10;}
            else   dvol=1;
    }

} 

void change_station(){
        audio.stopSong();
        astr=stations[dstation]; astr.trim();
        astr.toCharArray(buf, sizeof(buf));
        audio.connecttohost(buf);
        delay(500); // d
        log_i("free heap=%i", ESP.getFreeHeap());
}



void check_serial_radio(){
    if(Serial.available()){ //
        String r=Serial.readString(); r.trim();
        dstation=99;
        stations[dstation]=r;
      
    }
}

void check_vol_change(){
    // print() & println() can't handle printing long longs. (uint64_t)
    //this shows the received message from the clicker keyboard
    serialPrintUint64(results.value, HEX);
    Serial.println("");
    //clicker up and down arrows change the volume
    if (results.value==0xFF629D){
        dvol++;
        if (dvol>21){ dvol=1;}
    }
    if (results.value==0xFFA857){
        dvol--;
        if (dvol<1){ dvol=1;}
    }

    //if clicker # received toggle volume between very low and normal to give a mute option
        if (results.value==0xFF52AD){
            if (dvol==1){ dvol=10;}
            else   dvol=1;
        }
       
}

void check_station_change(){
 // now use IR << >> btns to change stations up or down
        if (results.value==0xFF22DD){
        dstation--;
        if (dstation<0){ dstation=maxstation-1;}
        }
        if (results.value==0xFFC23D){
        dstation++;
        if (dstation>=maxstation){ dstation=0;}
        }
        // change index position = btn# * 4
        if (results.value==0xFF4AB5){
                dstation=0;
        }
        if (results.value==0xFF6897){
                dstation=4;
        }
        if (results.value==0xFF9867){
                dstation=8;
        }
        if (results.value==0xFFB04F){
                dstation=12;
        }
        if (results.value==0xFF30CF){
                dstation=16;
        }
        if (results.value==0xFF18E7){
                dstation=20;
        }
        if (results.value==0xFF7A85){
                dstation=24;
        }
        if (results.value==0xFF10EF){
                dstation=28;
        }
        if (results.value==0xFF38C7){
                dstation=32;
        }
        if (results.value==0xFF5AA5){
                dstation=36;
        }
}

void check_MP3_play() {
     // asterisk * key on remote toggles between playing URLS or MP3s
        if (results.value==0xFF42BD){
            audio.stopSong();
            if (playMP3==0){
              playMP3=1;
              dstation=0;
              open_new_song(song_list[dstation]);
            }
            else{
              playMP3=0;
              dstation =0;
              change_station();
            }
            
            display.clear();
            Serial.printf("playMPS = %d \n", playMP3);
        }
        // ok or play button on remote
        if(results.value==0xFF02FD){
            
            
            Serial.println("Stopping or Pausing or Restarting");
            // audio.stopSong();
            if (audio.isRunning()==true) {
                audio.stopSong();
                audio.pauseResume();
                bstop=1;
            }
            else {
                bstop=0;
                audio.pauseResume();
                if(playMP3==1)open_new_song(song_list[dstation]);
                if(playMP3==0) change_station();
            }
        }



}

void open_new_song(String filename){
    audio.stopSong();
    filename="/mp3/" + filename;
    Serial.println(filename);
    music_info.name = filename.substring(1, filename.indexOf("."));
    audio.connecttoFS(SD_MMC, filename.c_str());
    music_info.runtime = audio.getAudioCurrentTime();
    music_info.length = audio.getAudioFileDuration();
    music_info.volume = audio.getVolume();
    music_info.status = 1;
    Serial.println("**********start a new sound************");
}

void display_music(){
    char buff[20];
    display.clear();
    //display.setFont(ArialMT_Plain_10);
    sprintf(buff, "Index %d vol %d", position,dvol);
    display.drawString(0, 0, buff );
    display.drawStringMaxWidth(0,25,128,music_info.name);
    display.display();

}
 
void logoshow(void){
    display.clear();
    display.drawStringMaxWidth(0,0,128, "Music Player");
    display.display();
    delay(2000);
}
 
void lcd_text(String text){
    display.clear();
    display_StationVolume();
    display.drawStringMaxWidth(0,20,128,text.c_str());
    display.display();
    delay(2000);
}

void display_StationVolume(){
    display.clear();
    sprintf(buf_sta_vol,"Station %d Vol` %d \n",dstation,dvol);
    display.drawStringMaxWidth(0,0,128,buf_sta_vol);
    display.drawStringMaxWidth(0, 25, 128, dinfo);
    display.display();
    //delay(500);
}
 
void print_song_info(){
    Serial.println("***********************************");
    Serial.println(audio.getFileSize());
    Serial.println(audio.getFilePos());
    Serial.println(audio.getSampleRate());
    Serial.println(audio.getBitsPerSample());
    Serial.println(audio.getChannels());
    Serial.println(audio.getVolume());
    Serial.println("***********************************");
}
 
void print_song_time(){
    //Serial.println(audio.getAudioCurrentTime());
    //Serial.println(audio.getAudioFileDuration());
    music_info.runtime = audio.getAudioCurrentTime();
    music_info.length = audio.getAudioFileDuration();
    music_info.volume = audio.getVolume();
}
 
// optional
void audio_info(const char *info){
    Serial.print("info        "); Serial.println(info);
    dinfo=info;
    display_StationVolume();
   
}

void audio_id3data(const char *info){  //id3 metadata
    Serial.print("id3data     ");Serial.println(info);
}

void audio_eof_mp3(const char *info){  //end of file
    Serial.print("eof_mp3     ");Serial.println(info);
}

void audio_showstation(const char *info){
    sprintf(buf_sta_vol,"Station %d Vol` %d \n",dstation,dvol);
    Serial.println(buf_sta_vol);      
    Serial.print("station     ");Serial.println(info);
    dinfo=info;
    display_StationVolume();
   
}

void audio_showstreamtitle(const char *info){
    Serial.print("streamtitle ");
    Serial.println(info);
    dinfo=info;
    display_StationVolume();
}

void audio_bitrate(const char *info){
    Serial.print("bitrate     ");Serial.println(info);
}

void audio_commercial(const char *info){  //duration in sec
    Serial.print("commercial  ");Serial.println(info);
}

void audio_icyurl(const char *info){  //homepage
    Serial.print("icyurl      ");Serial.println(info);
}

void audio_lasthost(const char *info){  //stream URL played
    Serial.print("lasthost    ");Serial.println(info);
}

int get_music_list(fs::FS &fs, const char *dirname, uint8_t levels, String wavlist[100]) {
    Serial.printf("Listing directory: %s\n", dirname);
    int i = 0;
    File root = fs.open(dirname);
    if (!root) {
        Serial.println("Failed to open directory");
        return i;
    }
    if (!root.isDirectory()) {
        Serial.println("Not a directory");
        return i;
    }
    File file = root.openNextFile();
    while (file) {
        if (file.isDirectory()) {
        }
        else {
            String temp = file.name();
            if (temp.endsWith(".wav")) {
                wavlist[i] = temp;
                i++;
            }
            else if (temp.endsWith(".mp3")) {
                wavlist[i] = temp;
                i++;
            }
        }
        file = root.openNextFile();
    }
    return i;
}


int get_station_list(fs::FS &fs, const char *dirname, uint8_t levels, String stations[100]) {
  File url_list = fs.open("/URLs/URL_LIST.txt");
  String str;
  int count = 0;
  Serial.println("\nStation urls:");
  while (url_list.available() && count < MAX_LINES) {
    str= url_list.readStringUntil('\n');
    stations[count]=str;
    Serial.println(str);
    count++;
  }
  url_list.close();
  Serial.print("MAX_Stations = ");
  Serial.println(count);
  return(count);
}

**Comment by waszee** [comment]
FYI - The reason for preferring https instead of http: is that I do development on a Windows 11 system and the default Windows security system kept giving my dangerous warning popups or refused connections in my browsers.  It is not a problem on my raspberry pi systems and agree with yoiur comment.  The overhead did not seem to be problem though as the audio version plays those streams fine and your player version plays most of my streams well.  I can send you the full 58 stream list that I am currently using if you want but was trying to keep messages short.

**Comment by waszee** [comment]
A better reference for ESP32-audioI2S is the following.  I think this might have been the source for code examples that I built from  but I had it from a complex twisted path so not sure.

#include "Audio.h" //see Schhreibfaul1 repository at github "https://github.com/schreibfaul1/ESP32-audioI2S"

The simple radio example at his site might be good to use for comparing station playback.  Audio.h seems to detect different types of audio streams and automatically switches the codec as needed so that might be the difference.  I think audiotool player actually sounds a little better on some streams but has trouble playing others.  Still working understanding the differences.
